src/cdomains/intDomain.ml

```diff
@@ -1941,79 +1958,111 @@ module Enums : IkindUnawareS = struct
         | _ -> x :: merge_sub xs b
       )
   (* let merge_sub x y = Set.(diff (of_list x) (of_list y) |> to_list) *)
-  let join = curry @@ function
-    | Inc x, Inc y -> Inc (merge_cup x y)
-    | Exc (x,r1), Exc (y,r2) -> Exc (merge_cap x y, R.join r1 r2)
-    | Exc (x,r), Inc y
-    | Inc y, Exc (x,r) -> Exc (merge_sub x y, if y = [] then r else R.join r (R.of_interval range_ikind (List.hd y, List.last y)))
-  let meet = curry @@ function
+  let join_ignore_ikind = curry @@ function
+  | Inc x, Inc y -> Inc (merge_cup x y)
+  | Exc (x,r1), Exc (y,r2) -> Exc (merge_cap x y, R.join r1 r2)
+  | Exc (x,r), Inc y
+  | Inc y, Exc (x,r) ->
+    let r = if y = []
+      then r
+      else
+        let (min_el_range, max_el_range) = Tuple2.mapn  (fun x -> R.of_interval range_ikind (Size.min_range_sign_agnostic x)) (List.hd y,  List.last y) in
+        let range = R.join min_el_range max_el_range in
+        R.join r range
+    in
+    Exc (merge_sub x y, r)
+
+  let join ikind = join_ignore_ikind
+
+  let meet ikind = curry @@ function
     | Inc x, Inc y -> Inc (merge_cap x y)
     | Exc (x,r1), Exc (y,r2) -> Exc (merge_cup x y, R.meet r1 r2)
     | Inc x, Exc (y,r)
     | Exc (y,r), Inc x -> Inc (merge_sub x y)
   (* let join x y = let r = join x y in print_endline @@ "join " ^ short 10 x ^ " " ^ short 10 y ^ " = " ^ short 10 r; r *)
   (* let meet x y = let r = meet x y in print_endline @@ "meet " ^ short 10 x ^ " " ^ short 10 y ^ " = " ^ short 10 r; r *)
 
-  let widen x y = join x y
-  let narrow x y = meet x y
+  let widen = join
+  let narrow = meet
 
-  let leq x y = join x y = y
+  let leq x y = (join_ignore_ikind x y) = y
 
   let max_elems () = get_int "ana.int.enums_max" (* maximum number of resulting elements before going to top *)
```

Comment:
This option seems unnecessary now that only singleton `Inc` arithmetic is possible. We shouldn't be limiting the number of enum values we can analyze, because only finitely many can exist in a given program as they must literally be present in the code.

Singleton `Inc` arithmetic doesn't seem to require this bound for termination either. For example, `for (int i = 0; true; i++);`:
1. Initially: `Inc [0]`
2. First iteration gives `Inc [0] + Inc [1] = Inc [1]`
3. First iteration join gives `Inc[0] join Inc [1] = Inc [0, 1]`
4. Second iteration gives `Inc [0, 1] + Inc [1] = top`
5. Second iteration join gives `Inc [0, 1] join top = top`
6. Third iteration gives `top + Inc [1] = top`
7. Third iteration join gives `top join top = top`, fixpoint reached.

Or is there still some case that does require the upper bound for termination?

Reply:
I removed the precise arithmetic for unary operations (unary minus, bitwise negations) on inclusion sets with multiple elements (2bc92b0e58156a058ccf0268444a1987eebc88a8), which then allows to remove the cardinality limit.

Reply:
Was there a problem with doing arithmetic on multiple elements?
The idea behind Enums was to not lose information on join/widen but on arithmetic operations (so it that it's precise for enums).
Non-termination can only arise due to new values from arithmetic operations - that's what the `ana.int.enums_max` was for.
What's the difference between this singleton `Inc` arithmetic and `ana.int.enums_max = 1`?